src/api/informer.rs

```diff
@@ -126,43 +118,98 @@ impl<K> Informer<K> where
         self
     }
 
-
     /// Run a single watch poll
     ///
     /// If this returns an error, it resets the resourceVersion.
     /// This is meant to be run continually and events are meant to be handled between.
-    /// If handling all the events is too time consuming, you probably need a queue.
-    pub async fn poll(&self) -> Result<()> {
+    /// poll returns a Stream so events can be handled asynchronously
+    pub async fn poll(&self) -> Result<impl Stream<Item = Result<WatchEvent<K>>>> {
         trace!("Watching {:?}", self.resource);
-        match self.single_watch().await {
-            Ok((events, newver)) => {
-                *self.version.write().unwrap() = newver;
-                for e in events {
-                    self.events.write().unwrap().push_back(e);
-                }
-            },
+
+        // Create our watch request
+        let req = self.resource.watch(&self.params, &self.version())?;
+
+        // Clone our version so we can move it into the Stream handling
+        // and avoid a 'static lifetime on self
+        let version = self.version.clone();
+
+        // Fetch an initial version as well to use for if we encounter
+        // a desync issue
+        let initial_version = self.get_resource_version().await?;
+
+        // Attempt to fetch our stream
+        let stream = self.client.request_events::<WatchEvent<K>>(req).await;
+
+        match stream {
+            Ok(events) => {
+                // Add a map stage to the stream which will update our version
+                // based on each incoming event
+                Ok(events
+                    .map(move |event| {
+                        // Check if we need to update our version based on the incoming events
+                        let new_version = match &event {
+                            Ok(WatchEvent::Added(o))
+                            | Ok(WatchEvent::Modified(o))
+                            | Ok(WatchEvent::Deleted(o)) => o.meta().resourceVersion.clone(),
+                            Ok(WatchEvent::Error(e)) => {
```

Comment:
It doesn't seem like we can repoll in the current stream, as we've lost access to it at this point in the map. Instead what I've done is added a new error type `KubeDesyncError`, and on a version mismatch we will return one of these. 

Then I've added a filter to the stream that will yield elements until the first KubeDesyncError (if any). That way a desync issue stops our stream short, and in the handling of it we also wait for 10s like `reset` does before allowing another poll.

Reply:
I'm not sure we need to convert the error itself into a DesyncError (because it could be another error), and maybe it's necessary for imformer users to handle the event themselves.

I do like resetting to what we saw before we started as a safeguard, but it is an extra http GET request. The data loss this prevents is meant to be fixed with [bookmarks](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks).

Maybe for now, we should just set a bool for `needs_rewatch = true` if we get a 410 GONE and avoid the extra call every poll?